static/modules/collision.js

```diff
@@ -115,39 +114,155 @@ export function isColliding(shapeA, shapeB) {
  * @returns {Vector}
  */
 export function collide(shapeA, shapeB, resolve = true) {
-  if (shapeA instanceof Box && shapeB instanceof Box) {
+  if (shapeA instanceof CollisionBox && shapeB instanceof CollisionBox) {
     return collide_BoxBox(
       /* @type {Box} */ (shapeA),
       /* @type {Box} */ (shapeB),
       resolve
     );
-  } else if (shapeA instanceof Circle && shapeB instanceof Box) {
+  } else if (
+    shapeA instanceof CollisionCircle &&
+    shapeB instanceof CollisionBox
+  ) {
     return collide_BoxCircle(
       /* @type {Box} */ (shapeB),
       /* @type {Circle} */ (shapeA),
       resolve
     );
-  } else if (shapeA instanceof Box && shapeB instanceof Circle) {
+  } else if (
+    shapeA instanceof CollisionBox &&
+    shapeB instanceof CollisionCircle
+  ) {
     return collide_BoxCircle(
       /* @type {Box} */ (shapeA),
       /* @type {Circle} */ (shapeB),
       resolve
     );
-  } else if (shapeA instanceof Circle && shapeB instanceof Circle) {
+  } else if (
+    shapeA instanceof CollisionCircle &&
+    shapeB instanceof CollisionCircle
+  ) {
     return collide_CircleCircle(
       /* @type {Circle} */ (shapeA),
       /* @type {Circle} */ (shapeB),
       resolve
     );
+  } else if (
+    shapeA instanceof CollisionCircle &&
+    shapeB instanceof CollisionBeam
+  ) {
+    return collide_CircleBeam(shapeA, shapeB, resolve);
+  } else if (
+    shapeA instanceof CollisionBeam &&
+    shapeB instanceof CollisionCircle
+  ) {
+    return collide_CircleBeam(shapeB, shapeA, resolve);
+  } else if (
+    shapeA instanceof CollisionBox &&
+    shapeB instanceof CollisionBeam
+  ) {
+    return collide_BoxBeam(shapeA, shapeB, resolve);
+  } else if (
+    shapeA instanceof CollisionBeam &&
+    shapeB instanceof CollisionBox
+  ) {
+    return collide_BoxBeam(shapeB, shapeA, resolve);
   } else {
     return new Vector(0, 0);
   }
 }
 
+/**
+ * Determines if a circle is colliding with a beam
+ * @param {CollisionCircle} circle
+ * @param {CollisionBeam} beam
+ * @param {boolean} resolve ignored. TODO implement
+ * @return {Vector}
+ */
+export function collide_CircleBeam(circle, beam, resolve) {
+  const dist = circle.pos.distToSeg(beam.p1, beam.p2);
+
+  let cVector = new Vector(0, 0);
+  if (dist < circle.radius + beam.thickness / 2) {
+    return new Vector(1, 1);
+  }
+  return cVector;
+}
+
+/**
+ * Determines if a box is colliding with a beam
+ * TODO this ignores beam thickness
```

Comment:
This TODO is a big one. I think treating it as a circle is actually better for now because of this TODO, and is much much faster.

Reply:
I realized it's actually easy, just run the same algorithm once for each side of the beam instead of once for the middle of the beam

Reply:
Implemented in 66a22a2